miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileDownloader.kt

```diff
@@ -0,0 +1,176 @@
+        callbackId: String,
+        fileName: String,
+        url: String,
+        headers: DownloadFileHeaderObj?,
+    ) {
+        val file = createFileDirectory(fileName = fileName)
+        val extension = url.substring(url.lastIndexOf("."))
+        val mimetype = mimeTypeMap.getMimeTypeFromExtension(extension)
+        val client = createHttpClient(headers)
+        val request = Request.Builder().url(url).build()
+        val response = client.newCall(request).execute()
+        if (response.isSuccessful) {
+            response.body?.let { responseBody ->
+                writeInputStreamToFile(inputStream = responseBody.byteStream(), file = file)
+                openShareIntent(mimetype, file)
+                bridgeExecutor.postValue(callbackId, fileName)
+            } ?: run {
+                bridgeExecutor.postError(callbackId, "$ERR_FILE_DOWNLOAD $ERR_EMPTY_RESPONSE_BODY")
+            }
+        } else {
+            bridgeExecutor.postError(callbackId, "$ERR_FILE_DOWNLOAD ${response.code}")
+        }
+    }
+
+    @VisibleForTesting
+    internal fun createHttpClient(headers: DownloadFileHeaderObj?): OkHttpClient {
+        val builder = OkHttpClient.Builder()
+        headers?.token?.let { token ->
+            builder.addNetworkInterceptor {
+                val requestBuilder = it.request().newBuilder()
+                requestBuilder.addHeader("token", token)
+                it.proceed(requestBuilder.build())
+            }
+        }
+        return builder.build()
+    }
+
+    @VisibleForTesting
+    internal fun createFileDirectory(fileName: String): File {
+        val cacheDir = File("${activity.cacheDir}/mini_app_download")
+        if (!cacheDir.exists()) {
+            cacheDir.mkdirs()
+        }
+        cache = cacheDir
+        return File(cacheDir, fileName)
+    }
+
+    @VisibleForTesting
+    @Suppress("NestedBlockDepth", "MagicNumber")
+    internal fun writeInputStreamToFile(inputStream: InputStream, file: File) = try {
+        inputStream.use { input ->
+            var size: Int
+            val buffer = ByteArray(2048)
+            FileOutputStream(file).use { fos ->
+                BufferedOutputStream(fos, buffer.size).use { bos ->
+                    while (input.read(buffer, 0, buffer.size)
+                            .also { size = it } != -1
+                    ) {
+                        bos.write(buffer, 0, size)
+                    }
+                    bos.flush()
+                }
+            }
+        }
+    } catch (e: IOException) {
+        Log.e(TAG, "Failed to write in the directory.", e)
+    }
+
+    @VisibleForTesting
+    internal fun openShareIntent(mimetype: String?, file: File) {
+        val intent = Intent(Intent.ACTION_SEND)
+        intent.type = mimetype
+        intent.putExtra(
+            Intent.EXTRA_STREAM,
+            DefaultFileProvider(activity).getUriForFile(file)
+        )
+        activity.startActivity(Intent.createChooser(intent, null))
```

Comment:
Sorry I think it was missing in the specs. This should use `ACTION_CREATE_DOCUMENT` instead of `ACTION_SEND`. Actually the `WebViewFileDownloader` should maybe be updated to use that as well.

This would actually simplify some things because you don't need to create a cache - you can just save directly to the file. However, this also means we need a way of receiving the Intent result from the Host App. Here's some example code:
```
internal fun openShareIntent(mimetype: String?, fileName: String) {
    val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
        addCategory(Intent.CATEGORY_OPENABLE)
        type = mimetype
        putExtra(Intent.EXTRA_TITLE, fileName)
    }
    activity.startActivityForResult(intent, MINI_APP_FILE_DOWNLOAD_REQUEST_CODE)
}

fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
    if (requestCode == MINI_APP_FILE_DOWNLOAD_REQUEST_CODE) {
        intent?.data?.let { destinationUri ->
            try {
                // Download file...

                activity.contentResolver.openOutputStream(destinationUri)?.let { outputStream ->
                    outputStream.write(responseBody.bytes())
                    outputStream.close()
                } ?: Log.e("Downloader","Failed to download file: could not open OutputStream.")
            } catch (e: IllegalArgumentException) {
                // Error
            }
        }
    }
}
```